samples/nrf9160/modem_shell/src/gnss/gnss_interface.c

```diff
@@ -456,28 +451,176 @@ static int gnss_enable_all_nmeas(void)
+		return err;
+	}
+
+	serving_cell->mnc = strtol(&resp_buf[3], NULL, 10);
+	/* Null-terminate MCC, read and store it. */
+	resp_buf[3] = '\0';
+	serving_cell->mcc = strtol(resp_buf, NULL, 10);
+
+	return 0;
+}
+#endif /* CONFIG_NRF_CLOUD_AGPS */
+
+#if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_SUPL_CLIENT_LIB)
 static void get_agps_data(struct k_work *item)
 {
 	ARG_UNUSED(item);
 
 	char flags_string[48];
+	struct nrf_modem_gnss_agps_data_frame agps_request = {0};
 
-	get_agps_data_flags_string(flags_string, agps_data.data_flags);
+	if (agps_inject_ephe) {
+		agps_request.sv_mask_ephe = agps_need.sv_mask_ephe;
+	}
+	if (agps_inject_alm) {
+		agps_request.sv_mask_alm = agps_need.sv_mask_alm;
+	}
+	if (agps_inject_utc) {
+		agps_request.data_flags |=
+			agps_need.data_flags & NRF_MODEM_GNSS_AGPS_GPS_UTC_REQUEST;
+	}
+	if (agps_inject_klob) {
+		agps_request.data_flags |=
+			agps_need.data_flags & NRF_MODEM_GNSS_AGPS_KLOBUCHAR_REQUEST;
+	}
+	if (agps_inject_neq) {
+		agps_request.data_flags |=
+			agps_need.data_flags & NRF_MODEM_GNSS_AGPS_NEQUICK_REQUEST;
+	}
+	if (agps_inject_time) {
+		agps_request.data_flags |=
+			agps_need.data_flags & NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST;
+	}
+	if (agps_inject_pos) {
+		agps_request.data_flags |=
+			agps_need.data_flags & NRF_MODEM_GNSS_AGPS_POSITION_REQUEST;
+	}
+	if (agps_inject_int) {
+		agps_request.data_flags |=
+			agps_need.data_flags & NRF_MODEM_GNSS_AGPS_INTEGRITY_REQUEST;
+	}
+
+	get_agps_data_flags_string(flags_string, agps_request.data_flags);
 
 	mosh_print(
-		"GNSS: Getting AGPS data (ephe: 0x%08x, alm: 0x%08x, flags: %s)...",
-		agps_data.sv_mask_ephe,
-		agps_data.sv_mask_alm,
+		"GNSS: Getting A-GPS data from %s (ephe: 0x%08x, alm: 0x%08x, flags: %s)...",
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+		"nRF Cloud",
+#else
+		"SUPL",
+#endif
+		agps_request.sv_mask_ephe,
+		agps_request.sv_mask_alm,
 		flags_string);
 
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+	int err;
+
+	err = nrf_cloud_jwt_generate(0, jwt_buf, sizeof(jwt_buf));
+	if (err) {
+		mosh_error("GNSS: Failed to generate JWT, error: %d", err);
+		return;
+	}
+
+	struct nrf_cloud_rest_context rest_ctx = {
+		.connect_socket = -1,
+		.keep_alive = false,
+		.timeout_ms = CONFIG_NRF_CLOUD_REST_RECV_TIMEOUT * MSEC_PER_SEC,
+		.auth = jwt_buf,
+		.rx_buf = rx_buf,
+		.rx_buf_len = sizeof(rx_buf),
+		.fragment_size = 0, /* Defaults to CONFIG_NRF_CLOUD_REST_FRAGMENT_SIZE when 0 */
+		.status = 0,
+		.response = NULL,
+		.response_len = 0,
+		.total_response_len = 0
+	};
+
+	struct nrf_cloud_rest_agps_request request = {
+		.type = NRF_CLOUD_REST_AGPS_REQ_CUSTOM,
+		.agps_req = &agps_request,
+		.net_info = NULL
+	};
+
+	struct nrf_cloud_rest_agps_result result = {
+		.buf = agps_data_buf,
+		.buf_sz = sizeof(agps_data_buf),
+		.agps_sz = 0
+	};
+
+	struct lte_lc_cells_info net_info = { 0 };
+
+	err = serving_cell_info_get(&net_info.current_cell);
+	if (err) {
+		mosh_warn("GNSS: Could not get cell info, error: %d", err);
```

Comment:
was wondering that is it worth to continue? net_info is optional but cloud could not return any assistance data without this, right?

Reply:
Yes it can, cell information is only needed for position assistance. If cell information is not available, cloud will return all other assistance data.